### PR TITLE
Add aws_cognito_user resource

### LINKED ISSUE #4542: Add aws_cognito_user resource
### Description

Currently the aws_cognito has an `aws_cognito_user_group` resource which represents a group of users. In the AWS IDP console there is an option to create a user, and assign it to groups. Currently terraform does not support it and it is needed to be created manually.

### New or Affected Resource(s)

* aws_cognito_user

### References

image

### PR DESCRIPTION
**Pull request by rtim75**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #4542

Output from acceptance testing:

```
make testacc PKG_NAME=internal/service/cognitoidp TESTARGS='-run=TestAccCognitoUser'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cognitoidp/... -v -count 1 -parallel 20 -run=TestAccCognitoUser -timeout 180m
=== RUN   TestAccCognitoUser_basic
=== PAUSE TestAccCognitoUser_basic
=== RUN   TestAccCognitoUser_disappears
=== PAUSE TestAccCognitoUser_disappears
=== RUN   TestAccCognitoUser_temporaryPassword
=== PAUSE TestAccCognitoUser_temporaryPassword
=== RUN   TestAccCognitoUser_password
=== PAUSE TestAccCognitoUser_password
=== RUN   TestAccCognitoUser_attributes
=== PAUSE TestAccCognitoUser_attributes
=== RUN   TestAccCognitoUser_enabled
=== PAUSE TestAccCognitoUser_enabled
=== CONT  TestAccCognitoUser_basic
=== CONT  TestAccCognitoUser_attributes
=== CONT  TestAccCognitoUser_password
=== CONT  TestAccCognitoUser_disappears
=== CONT  TestAccCognitoUser_enabled
=== CONT  TestAccCognitoUser_temporaryPassword
--- PASS: TestAccCognitoUser_disappears (51.18s)
--- PASS: TestAccCognitoUser_basic (61.03s)
--- PASS: TestAccCognitoUser_enabled (99.19s)
--- PASS: TestAccCognitoUser_attributes (99.97s)
--- PASS: TestAccCognitoUser_temporaryPassword (137.58s)
--- PASS: TestAccCognitoUser_password (138.89s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cognitoidp 140.789s
```

### CODE REVIEW

File: aws/resource_aws_cognito_user.go
Comment (rtim75) [inline aws/resource_aws_cognito_user.go:216]: I'm not really sure if this is the best way to remove `sub` attribute from the resource. 

From [aws docs](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html):
> The SignUp API generates a persistent UUID for your user, and uses it as the immutable username attribute internally. This UUID has the same value as the sub claim in the user identity token.

So every user has this attribute but it cannot be modified or deleted.

**Comment by rtim75** [comment]
Another thing I'm not sure about is `client_metadata` attribute in the [inputs](https://github.com/aws/aws-sdk-go/blob/v1.38.64/service/cognitoidentityprovider/api.go#L12233-L12360). 

I'm thinking about adding it with `ForceNew` attribute.

File: aws/resource_aws_cognito_user.go
Comment (rtim75) [inline aws/resource_aws_cognito_user.go:39]: https://github.com/aws/aws-sdk-go/blob/v1.38.64/service/cognitoidentityprovider/api.go#L15444-L15454

I assume the field should be marked sensitive here as well then

**Comment by tomelliot16** [comment]
Any status update on this?

**Comment by rtim75** [comment]
> Any status update on this?

Hey, as you probably have noticed I've been working very slowly on this one. I will try to finish at least the main resource(`aws_cognito_user`) in the near future. From there things should go quicker.

**Comment by rtim75** [comment]
Please, let me know if this 

```
"mfa_preference": {
            Type: schema.TypeSet,
            Elem: &schema.Resource{
	            Schema: map[string]*schema.Schema{
		            "sms_enabled": {
			            Type: schema.TypeBool,
			            Optional: true,
            
		            },
		            "software_token_enabled": {
			            Type: schema.TypeBool,
			            Optional: true,
		            },
		            "preferred_mfa": {
			            Type: schema.TypeString,
			            Optional: true,
			            ValidateFunc: validation.StringInSlice([]string{
				            cognitoidentityprovider.ChallengeNameTypeSmsMfa,
				            cognitoidentityprovider.ChallengeNameTypeSoftwareTokenMfa,
			            }, false),
		            },
	            },
            },
            Optional: true,
            MaxItems: 1,
            },
```

makes sense for the [MfaPreference](https://docs.aws.amazon.com/sdk-for-go/api/service/cognitoidentityprovider/#AdminSetUserMFAPreferenceInput). I tried to avoid Resources inside of Resources and it looks nice to me. 
For constants I couldn't find a better match then `SMS_MFA` and `SOFTWARE_TOKEN_MFA`.

Expand would look like this(I didn't test this):
```
smsEnabled := mfaPreference["sms_enabled"].(bool)
softwareTokenEnabled := mfaPreference["software_token_enabled"].(bool)
preferredMfa := mfaPreference["preferred_mfa"].(string)

smsPreferred := preferredMfa == cognitoidentityprovider.ChallengeNameTypeSmsMfa
softwareTokenPreferred := preferredMfa == cognitoidentityprovider.ChallengeNameTypeSoftwareTokenMfa

return &cognitoidentityprovider.SMSMfaSettingsType{
	Enabled: aws.Bool(smsEnabled),
	PreferredMfa: aws.Bool(smsPreffered)
}, &cognitoidentityprovider.SoftwareTokenMfaSettingsType{
	Enabled: aws.Bool(softwareTokenEnabled),
	PreferredMfa: aws.Bool(softwareTokenPreferred),
}
```

**Comment by zhelding** [comment]
Pull request #21306 has significantly refactored the AWS Provider codebase. As a result, most PRs opened prior to the refactor now have merge conflicts that must be resolved before proceeding.

Specifically, PR #21306 relocated the code for all AWS resources and data sources from a single `aws` directory to a large number of separate directories in `internal/service`, each corresponding to a particular AWS service. This separation of code has also allowed for us to simplify the names of underlying functions -- while still avoiding namespace collisions.

We recognize that many pull requests have been open for some time without yet being addressed by our maintainers. Therefore, we want to make it clear that resolving these conflicts _in no way affects the prioritization_ of a particular pull request. Once a pull request has been prioritized for review, the necessary changes will be made by a maintainer -- either directly or in collaboration with the pull request author.

For a more complete description of this refactor, including examples of how old filepaths and function names correspond to their new counterparts: please refer to issue #20000.

For a quick guide on how to amend your pull request to resolve the merge conflicts resulting from this refactor and bring it in line with our new code patterns: please refer to our [Service Package Refactor Pull Request Guide](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/service-package-pullrequest-guide.md).

**Comment by breathingdust** [comment]
Hi all 👋  Just letting you know that this is issue is featured on this quarters roadmap. If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Comment by rtim75** [comment]
so, according to [TOTP Software Token MFA](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-mfa-totp.html):  
> Configuring TOTP for your user is a multi-step process where your user receives a secret code that they validate by entering a one-time password. Next, you can enable TOTP MFA for your user or set TOTP as the preferred MFA method for your user.

which means the preferred MFA method cannot be set at user creation, so I'm going to add only `Computed` field to export this information if it's present

File: internal/service/cognitoidp/user.go
Comment (rtim75) [inline internal/service/cognitoidp/user.go:92]: I'm not sure how to make the `attributes` to be a block. According to the [Plugin Development guide](https://www.terraform.io/docs/extend/schemas/schema-types.html#typemap) next schema

```go
"tags": {
  Type:     schema.TypeMap,
  Elem: &schema.Schema{
    Type: schema.TypeString,
  },
},
```
should result in block argument:
```hcl
resource "example_compute_instance" "ex" {
  tags {
    env = "development"
    name = "example tag"
  }
}
```

But I'm getting attribute configuration:


```hcl
attributes = {
    bar            = "baz"
}
```

File: internal/service/cognitoidp/user.go
Comment (rtim75) [inline internal/service/cognitoidp/user.go:92]: I'm not sure how to make the `attributes` to be a block. According to the [Plugin Development guide](https://www.terraform.io/docs/extend/schemas/schema-types.html#typemap) next schema

```go
"tags": {
  Type:     schema.TypeMap,
  Elem: &schema.Schema{
    Type: schema.TypeString,
  },
},
```
should result in block argument:
```hcl
resource "example_compute_instance" "ex" {
  tags {
    env = "development"
    name = "example tag"
  }
}
```

But I'm getting attribute configuration:


```hcl
attributes = {
    bar            = "baz"
}
```

**Comment by rtim75** [comment]
While I'm working on attributes tests, I'd like to ask what should I do with client-side attributes the resource supports and I believe should include, like `client_metadata` for example. I'm not aware of any resource that has such functionality so I'm not sure what to do in terms of testing those.

---

Also, I just realised that #4542 talks about including user into user group. Looking at IAM this should be implemented as a separate resource. Should I create an accompanying PR with that resource or I can include it into this one? From the [contributing guides](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/contribution-checklists.md#new-resource) I assume it should be a separate PR.

**Comment by rtim75** [comment]
I can't get my head around this: after (temporary_)password is deleted it's saved as an empty string in the state file rather than `null` which it is at the resource creation. 

```
14:57:38 ❯ make testacc PKG_NAME=internal/service/cognitoidp TESTARGS='-run=TestAccCognitoUser_.*sword'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cognitoidp/... -v -count 1 -parallel 20 -run=TestAccCognitoUser_.*sword -timeout 180m
=== RUN   TestAccCognitoUser_temporaryPassword
=== PAUSE TestAccCognitoUser_temporaryPassword
=== RUN   TestAccCognitoUser_password
=== PAUSE TestAccCognitoUser_password
=== CONT  TestAccCognitoUser_temporaryPassword
=== CONT  TestAccCognitoUser_password
=== CONT  TestAccCognitoUser_temporaryPassword
    user_test.go:93: Step 4/4 error: Check failed: Check 2/3 error: aws_cognito_user.test: Attribute 'temporary_password' found when not expected
=== CONT  TestAccCognitoUser_password
    user_test.go:149: Step 4/4 error: Check failed: Check 2/3 error: aws_cognito_user.test: Attribute 'password' found when not expected
--- FAIL: TestAccCognitoUser_temporaryPassword (111.11s)
--- FAIL: TestAccCognitoUser_password (113.51s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/cognitoidp 115.433s
FAIL
make: *** [testacc] Error 1
```

**Comment by rtim75** [comment]
> A possible solution could be setting default value for both password as an empty string.

so this didn't help at all

> I would update tests to check for empty strings if this issue weren't causing changes to outputs:

I see at least user pool works the same way so I guess I will update tests to check for empty string and mark the PR as ready.

File: internal/service/cognitoidp/user.go
Comment (gdavison) [inline internal/service/cognitoidp/user.go:89]: We prefer matching the AWS API more closely. I would suggest replacing this with top-level parameters `mfa_setting_list` of type `schema.TypeSet` and `preferred_mfa_setting` of type `schema.TypeString`.

Since `MFAOptions` is no longer supported according to AWS API documentation, it can be ignored.

File: internal/service/cognitoidp/user.go
Comment (gdavison) [inline internal/service/cognitoidp/user.go:368]: Clearing both `password` and `temporary_password` should probably call `AdminResetUserPassword` to clear the user's password

**Review by gdavison — CHANGES_REQUESTED** [review]
Thanks for the PR, @rtim75. I have a number of suggestions for changes to make. Please also check the results of the automated checks

**Comment by rtim75** [comment]
thank you @gdavison for your review. I will take a look at your comments and fix everything this weekend. Do you have have any guidelines regarding implementation and testing of client-side arguments? Cognito user api actions have a bunch of such arguments and I believe they are pretty crucial:

- desired delivery mediums
- message action
- client metadata
etc.

--- 
Also, I removed computed timestamps according to [common review items](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/pullrequest-submission-and-lifecycle.md#common-review-items) but I can see that most if not all the resources from `cognitoidp` have those. Should I bring them back for consistency or leave them skipped?

**Comment by gdavison** [comment]
Testing of the client-side arguments can get challenging, since many of them are handled by a Lambda function that's a part of an authentication/authorization system, and it could be implementation-dependent. For the most part, we rely on the AWS API catching major errors.

For implementation in Terraform, in general we try to match the AWS API structures. Sometimes this can end up looking a little strange, like the MFA information. However, it makes dealing with changes on the AWS side easier, for example if they added another MFA contact method, they could just add an item to the list, instead of needing to add another Boolean parameter.

**Comment by gdavison** [comment]
For the timestamp fields, I would include them for consistency with the other resources.

File: internal/service/cognitoidp/user.go
Comment (rtim75) [reply internal/service/cognitoidp/user.go:350]: I like your idea but it comes with additional complications: `AdminResetUserPassword` action requires users to have `email` or `phone_number` attributes set(depends on user pool configuration) when they are in the `CONFIRMED` state. And simply doesn't allow resetting password when in `FORCE_CHANGE_PASSWORD` state:

> FORCE_CHANGE_PASSWORD
```
Error: Error resetting Cognito User's password (user_pool/example): NotAuthorizedException: User password cannot be reset in the current state.
```

> CONFIRMED and no attributes
```
Error: Error resetting Cognito User's password (user_pool/example): InvalidParameterException: Cannot reset password for the user as there is no registered/verified email or phone_number
```

Are we ok with forcing tf users to set `email`/`phone_number` in case they decide to remove passwords? Also I'm not sure how to handle `User password cannot be reset in the current state` error. We somehow need to change the users's state.

File: internal/service/cognitoidp/user.go
Comment (rtim75) [inline internal/service/cognitoidp/user.go:273]: When I was testing, I noticed that `AdminUpdateUserAttributesInput` also has a `ClientMetadata` field. So I believe it should be added here as well.

File: internal/service/cognitoidp/user.go
Comment (rtim75) [reply internal/service/cognitoidp/user.go:359]: @gdavison could you provide your thoughts on this, please?

Comment (gdavison) [reply internal/service/cognitoidp/user.go:359]: I that case, we should leave the code as-is. It's probably worth adding a note to the documentation that clearing the `temporary_password` or `password` does not remove the password for the user

**Comment by rtim75** [comment]
@gdavison The PR is ready for the second review. All the discussions have been resolved. The only failing action is caused by resources unrelated to this PR.

**Comment by cemo** [comment]
Gentle ping @gdavison :) We desperately need this PR

**Review by gdavison — APPROVED** [review]
Looks good! 🚀 

--- PASS: TestAccCognitoUser_disappears (40.50s)
--- PASS: TestAccCognitoUser_basic (46.74s)
--- PASS: TestAccCognitoUser_enabled (68.43s)
--- PASS: TestAccCognitoUser_attributes (69.27s)
--- PASS: TestAccCognitoUser_password (86.09s)
--- PASS: TestAccCognitoUser_temporaryPassword (86.16s)

**Comment by pjroth** [comment]
Thanks so much @rtim75 for adding this!  I'm looking forward to being able to use this soon!

**Pull request merged**
